### ISSUE TITLE
Discrepancy between reported series and series present in the block

### ISSUE THREAD

**Issue by wiardvanrij** [prometheus/prometheus#13004]
### What did you do?

We had an OOM kill and to my surprise we basically duplicated in the amount of series. See screenshot: 
![image](https://github.com/prometheus/prometheus/assets/5786097/acf4dcac-57d2-406a-8e59-098e8c26288b)

I was investigating it, and obviously when turning around some stones I found some things we could improve on. However, when I dug further, checking the blocks themselves, I notice a different number in the amount of series:

```
promtool tsdb list .
BLOCK ULID                  MIN TIME       MAX TIME       DURATION     NUM SAMPLES  NUM CHUNKS   NUM SERIES   SIZE
01HCT4FHKH8CFEDS0VDRJCYPJY  1697378400000  1697385600000  2h0m0s       4440544312   37409715     10168069     7768846944
01HCTBB8VPR9CNZPGXVXPG2956  1697385600000  1697392800000  2h0m0s       4470131819   37641991     10216217     7825614680
01HCTJ703HY5QYW9SM1C2WM456  1697392800000  1697400000000  2h0m0s       4475921724   37603353     10122896     7821507631
01HCTS2QBN3ZVJDSYDBD687TZR  1697400000000  1697407200000  2h0m0s       4482466962   37737103     10214242     7802420218
```

So, I'm seeing:

- UI /tsdb-status reports:  Number of Series = 18922003
- prometheus_tsdb_head_series metric, reports roughly the same ^ 18m
- `count({__name__=~".+"})` = 8904164 (almost half)
- tsdb list, see above (they are somewhat older blocks, but are somewhat equal to the reported count query)

### What did you expect to see?

That the metric & the API would report roughly the same amount of series as what the blocks reports

### What did you see instead? Under which circumstances?

...

### System information

Docker image

### Prometheus version

```text
v2.44.0
```

### Prometheus configuration file

```yaml
args


      --web.console.templates=/etc/prometheus/consoles
      --web.console.libraries=/etc/prometheus/console_libraries
      --config.file=/etc/prometheus/config_out/prometheus.env.yaml
      --web.enable-lifecycle
      --web.enable-remote-write-receiver
      --enable-feature=memory-snapshot-on-shutdown
      --web.external-url=https://redacted/prometheus
      --web.route-prefix=/prometheus
      --storage.tsdb.wal-compression
      --storage.tsdb.retention.time=30d
      --storage.tsdb.retention.size=224000MB
      --storage.tsdb.path=/prometheus
      --web.enable-admin-api
      --web.config.file=/etc/prometheus/web_config/web-config.yaml
      --storage.tsdb.max-block-duration=2h
      --storage.tsdb.min-block-duration=2h
```
```


### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

**Comment by prymitive** [comment]
`count({__name__=~".+"})` will only count time series with samples from the past 5 minutes.
Seems like you have 50% churn rate, check `scrape_series_added` to see if you keep adding time series all the time.
If this is k8s environment then that's not unusual - k8s exports a lot of ephemeral label values like pod IDs etc - these keep changing as pods are restarted creating new time series.

**Comment by wiardvanrij** [comment]
I would agree with you, but please look at the screenshot after the restart. That does not make sense right? How can I out of the blue double the amount of series?

Furthermore, IF that would be the case, the amount of series in the block itself should be higher than what the count would provide, which is not the case either?

Anyhow, the **rate**/s on scrape_series_added is about ~50-150 and over time about calculated with an `increase[2h]` of ~800k.

**Comment by bboreham** [comment]
I thought about this for a bit, but couldn't figure out a way for it to go wrong.

> UI /tsdb-status reports: Number of Series = 18922003

This page has a number of break-downs; see if you can spot something that adds to the picture of where it went wrong.

Did it persist over a subsequent restart?

**Comment by wiardvanrij** [comment]
> Did it persist over a subsequent restart?

Yes, it did persist. See screenshot:
<img width="1319" alt="image" src="https://github.com/prometheus/prometheus/assets/5786097/65dea50e-0959-48d9-a7cd-609e0a3a1361">

>  This page has a number of break-downs; see if you can spot something that adds to the picture of where it went wrong.

Do you mean the break-downs like; Top 10 series count by metric names?  
If so, I actually analysed a few blocks for something else, which was the reason I 'found' this behaviour. Because the blocks themself in question report a different amount of time series. I don't see anything that can explain this behaviour and I'm also still a bit stumbled on why this is happening. 

I did a `promtool tsdb list .`  and that reports:

```
BLOCK ULID                  MIN TIME       MAX TIME       DURATION     NUM SAMPLES  NUM CHUNKS   NUM SERIES   SIZE
01HD9SQJVMP9R8Y1G9HA65P7HY  1697904000000  1697911200000  2h0m0s       4291006649   36154771     9825565      8046503793
01HDA0KA3MX4DGY4QT3JXWVABY  1697911200000  1697918400000  2h0m0s       4297123683   36203051     9844633      8004228397
01HDA7F1BPBP6QE0TS688R384P  1697918400000  1697925600000  2h0m0s       4304427474   36242132     9827859      8055649736
01HDAEARKMN476X64JGSN58RP6  1697925600000  1697932800000  2h0m0s       4311984338   36170351     9728502      8031849488
01HDAN6FVMHM788XMGX1RGKNFW  1697932800000  1697940000000  2h0m0s       4347782020   36540518     9946857      8118224219
01HDAW273MRF7MBT98VNCZWEKB  1697940000000  1697947200000  2h0m0s       4345390692   36454143     9793077      8104293826
```

**Comment by bboreham** [comment]
> Do you mean the break-downs like; Top 10 series count by metric names?

"Top 10 series count by metric names", "Top 10 series count by label value pairs"

**Comment by bboreham** [comment]
My suggestion was for you to think about the numbers, see if any of them seem out of line.

Note you can get more than 10 via the API, e.g. `http://my-prometheus:9090/api/v1/status/tsdb?limit=20`
(there is a bug that it caches the data for 30 seconds even when you change the limit)

**Comment by roidelapluie** [comment]
What does the Prometheus build info tell? Was it the same version before?

**Comment by wiardvanrij** [comment]
> What does the Prometheus build info tell? Was it the same version before?

Yes, it was not an upgrade. We "merely" OOM'd which caused the restart.

--

Also did not really find/see an indicator on the tsdb API if I'm honest.

**Comment by wiardvanrij** [comment]
Not forgotten, just haven't got the time yet to dig into this further.
